content/posts/pandastutor.md

```diff
@@ -0,0 +1,323 @@
+could somehow *serialize a binary memory snapshot* taken right after
+everything gets initialized. Then the user's browser can directly load
+that snapshot on the first page visit and cache it for future visits.
+(But maybe that snapshot would be too big to download efficiently, so
+nevermind!)
+
+
+### Bonus: other fine-tuning and ongoing work
+
+- [Graceful
+  degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation):
+  To continue supporting older browsers, our code detects if there are
+  any problems loading the Pyodide version of Pandas Tutor and, if so,
+  automatically reverts back to the original server-side version. Also,
+  [loadPyodide](https://pyodide.org/en/stable/usage/api/js-api.html#globalThis.loadPyodide)
+  crashes on some older mobile devices (presumably with [limited
+  memory](https://github.com/pyodide/pyodide/issues/533)?) and we can't
+  seem to catch the exception to recover; so we disable it by default on
+  mobile and add a toggle to let the user explicitly enable it. (This is
+  also nice since it prevents the page from preemptively downloading
+  large amounts of data on users' mobile phone plans.)
+- Auto-importer: We want users to be able to import any package when
+  writing their code in Pandas Tutor. But Pyodide's
+  [loadPackagesFromImports](https://pyodide.org/en/stable/usage/api/js-api.html#pyodide.loadPackagesFromImports)
+  works only on packages that come with Pyodide. To give users more
+  flexibility, we wrote an auto-importer that automatically calls
+  `micropip.install()` whenever the user's code tries to `import` an
+  unknown package. These package wheels get downloaded from PyPI
+  on-demand and cached in the browser. (Note that this works only if the
+  package name on PyPI matches the name in the `import` statement, but
+  this is often true.)
+- Preloading: As soon as the user visits the
+  [pandastutor.com](https://pandastutor.com/) landing page, it starts
+  downloading Pyodide and selected packages (e.g., pandas, numpy) in the
+  background. That way, by the time the user goes to the [visualizer
+  page](https://pandastutor.com/vis.html), these
+  packages may already have been cached in their browser, so they can
+  start working right away. (Thanks to Michael Kennedy from the [Talk
+  Python
+  podcast](https://talkpython.fm/episodes/show/358/understanding-pandas-visually-with-pandastutor)
+  for this suggestion!)
+- Caching file downloads: *[work-in-progress]* Some pandas code uses
+  functions like
+  [read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)
+  or
+  [read_html](https://pandas.pydata.org/docs/reference/api/pandas.read_html.html)
+  to download data files from the web. It would be great to cache those
+  files locally in-browser so that when the user re-runs that code in
+  the future, Pandas Tutor doesn't need to repeat the same
+  (potentially-large) downloads.
```

Comment:
I think if the remote server where the data is downloaded sets the cache headers right, they should already be cached, since we are using fetch requests and these are cached by the browser.

Though yes, maybe manual caching would still be preferable.